nighthawk/source/common/statistic_impl.cc

```diff
@@ -0,0 +1,214 @@
+      pow(a.mean() - b.mean(), 2) * a.count() * b.count() / combined->count();
+  return combined;
+}
+
+InMemoryStatistic::InMemoryStatistic() : streaming_stats_(std::make_unique<StreamingStatistic>()) {}
+
+void InMemoryStatistic::addValue(int64_t sample_value) {
+  samples_.push_back(sample_value);
+  streaming_stats_->addValue(sample_value);
+}
+
+uint64_t InMemoryStatistic::count() const {
+  ASSERT(streaming_stats_->count() == samples_.size());
+  return streaming_stats_->count();
+}
+double InMemoryStatistic::mean() const { return streaming_stats_->mean(); }
+double InMemoryStatistic::pvariance() const { return streaming_stats_->pvariance(); }
+double InMemoryStatistic::pstdev() const { return streaming_stats_->pstdev(); }
+
+std::unique_ptr<Statistic> InMemoryStatistic::combine(const Statistic& statistic) {
+  auto combined = std::make_unique<InMemoryStatistic>();
+  const InMemoryStatistic& b = dynamic_cast<const InMemoryStatistic&>(statistic);
+
+  combined->samples_.insert(combined->samples_.end(), this->samples_.begin(), this->samples_.end());
+  combined->samples_.insert(combined->samples_.end(), b.samples_.begin(), b.samples_.end());
+  combined->streaming_stats_ = this->streaming_stats_->combine(*b.streaming_stats_);
+  return combined;
+}
+
+const int HdrStatistic::SignificantDigits = 4;
+
+HdrStatistic::HdrStatistic() : histogram_(nullptr) {
+  // Upper bound of 60 seconds (tracking in nanoseconds).
+  const uint64_t max_latency = 1000L * 1000 * 1000 * 60;
+
+  int status = hdr_init(1 /* min trackable value */, max_latency, HdrStatistic::SignificantDigits,
+                        &histogram_);
+  if (status != 0) {
+    ENVOY_LOG(error, "Failed to initialize HdrHistogram.");
+    histogram_ = nullptr;
+  }
+}
+
+// TODO(oschaaf): valgrind complains when a Histogram is created but never used.
+HdrStatistic::~HdrStatistic() {
+  if (histogram_ != nullptr) {
+    hdr_close(histogram_);
+    histogram_ = nullptr;
+  }
+}
+
+void HdrStatistic::addValue(int64_t value) {
+  if (histogram_ != nullptr) {
+    // Failure to record a value can happen when it exceeds the configured minimum
+    // or maximum value we passed when initializing histogram_.
+    if (!hdr_record_value(histogram_, value)) {
+      ENVOY_LOG(warn, "Failed to record value into HdrHistogram.");
+    }
+  }
+}
+
+uint64_t HdrStatistic::count() const { return histogram_->total_count; }
+double HdrStatistic::mean() const { return hdr_mean(histogram_); }
+double HdrStatistic::pvariance() const {
+  return pstdev() * pstdev();
+  ;
+}
+double HdrStatistic::pstdev() const {
+  if (histogram_ == nullptr) {
+    return 0;
+  }
+  return hdr_stddev(histogram_);
+}
+
+std::unique_ptr<Statistic> HdrStatistic::combine(const Statistic& statistic) {
+  auto combined = std::make_unique<HdrStatistic>();
+  const HdrStatistic& b = dynamic_cast<const HdrStatistic&>(statistic);
+
+  if (this->histogram_ == nullptr || b.histogram_ == nullptr) {
+    return combined;
```

Comment:
I think it would majorly simplify if we just `throw` in the constructor when `nullptr` happens, so we don't need special case handling around the place.